### ISSUE TITLE
Added example cpp20_streams, which reproduces an assertion.

### ISSUE THREAD

**Issue by bveldhoen** [boostorg/redis#85]
I've created both a PR and an issue as I wasn't sure what would be the preferred way to deal with this. Feel free to close either one.

Please see this PR for a description:
https://github.com/boostorg/redis/pull/84

**Comment by mzimbres** [comment]
Hi, thanks for reporting, I will have a look. Does the problem persist if you use only one thread.

**Comment by bveldhoen** [comment]
Indeed, I also can't reproduce using only 1 thread.

**Comment by bveldhoen** [comment]
I've just added a commit to clean it up a bit.

**Comment by bveldhoen** [comment]
The current workaround I'm using is by giving the redis_client its own ioc (instead of sharing it with other classes that use it), and that ioc runs on a single thread. However, this shouldn't be a requirement for using boost::redis, correct? In other words, redis_client and boost::redis should work with an ioc that runs on multiple threads, even though the strand ensures that calls to async_exec and its completion handler aren't executed concurrently.

**Comment by mzimbres** [comment]
Correct, it shoud be possible to use boost.redis with multi-threaded io_contexts as well. There is some concurrent issue there that I couldn't spot yet, probably in the use of strand.

**Comment by mzimbres** [comment]
I have rewritten your example using some facility functions I am developing in another branch and couldn't also see any crash: https://gist.github.com/mzimbres/b4e25f4b95a1c4f5c8ce8aa29dde6352.

**Comment by bveldhoen** [comment]
I've added a new branch with some changes to the examples, see here:
https://github.com/bveldhoen/redis/tree/83-fix-reconnect-loop-in-the-subscriber-example-with-streams-example

Points of attention:

* I've had to add a call to req_.clear() in cpp17_streams. Without it, the example didn't function properly on my machine (it took a long time before the writes to the console became visible).
* The problem seemed to go away in cpp20_streams when using the new redis::async_run method. Other changes (i.e. members for req and resp) didn't make any difference. See the changes in the top commit for the fix for cpp20_streams.
* Going forward, I think the name cpp17_streams and the changed implementation are better suitable. cpp20_streams can be removed.
